### ISSUE TITLE
reduce method(decode)'s bytecode size

### ISSUE THREAD

**Issue by hellyguo** [lettuce-io/lettuce-core#1323]
#### Describe the solution you'd like

the method, `decode()`, in `RedisStateMachine` is so big. When convert it to bytecode, the size is great than 325 bytes.

#### Describe alternatives you've considered

using enum with behavior + lambda, we can reduce the method's size.

#### Teachability, Documentation, Adoption, Migration Strategy

[JitWatch](https://github.com/AdoptOpenJDK/jitwatch/wiki/JarScan)

**Comment by hellyguo** [comment]
before
```shell
➜  jitwatch-1.3.0 ./jarScan.sh --mode=maxMethodSize --limit=325 --packages=io.lettuce.core.protocol ~/code/github/lettuce-core/target/classes
"io.lettuce.core.protocol","RedisStateMachine","decode","io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",571
```

after
```shell
➜  jitwatch-1.3.0 ./jarScan.sh --mode=maxMethodSize --limit=5 --packages=io.lettuce.core.protocol ~/code/github/lettuce-core/target/classes|grep -F RedisStateMachine
"io.lettuce.core.protocol","RedisStateMachine","loopPeekStack","io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",127
"io.lettuce.core.protocol","RedisStateMachine","decode","io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",109
"io.lettuce.core.protocol","RedisStateMachine","funcMulti","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",100
"io.lettuce.core.protocol","RedisStateMachine","funcBulk","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",75
"io.lettuce.core.protocol","RedisStateMachine","funcInteger","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",42
"io.lettuce.core.protocol","RedisStateMachine","funcSingle","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",39
"io.lettuce.core.protocol","RedisStateMachine","funcBytes","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",32
"io.lettuce.core.protocol","RedisStateMachine","funcError","io.lettuce.core.protocol.RedisStateMachine,io.lettuce.core.protocol.RedisStateMachine$State,io.netty.buffer.ByteBuf,io.lettuce.core.protocol.RedisCommand<?, ?, ?>,io.lettuce.core.output.CommandOutput<?, ?, ?>",28
```

**Comment by mp911de** [comment]
Can you explain the motivation behind this change?

**Comment by hellyguo** [comment]
I'm working on a business project now. In my project, I need a fast 3rd lib to access redis. I may use some asynchronous scenarios. What `lettuce` has is What I needed.

I compared `jedis` and `lettuce`. After a simple(may be wrong) test, I found `lettuce` is not the fast one in my test.

I captured the thread dump. Found `RedisStateMachine`. Then I read the code, found the method.

My motivation is let `lettuce` run faster. But just a tiny improve.

- reduce the method(decode), more readable, less bytecode size
- split in several methods, may some is tiny enough that can be inline
- prefer to use enum-interface-lambda, not switch-case

[like what netty did](https://github.com/netty/netty/pull/5801)
